src/aruku/walking/process/kinematic.cpp

```diff
@@ -0,0 +1,682 @@
+  m_ssp_time_end_l(0), m_ssp_time_start_r(0), m_ssp_time_End_r(0), m_phase_time1(0),
+  m_phase_time2(0), m_phase_time3(0), m_x_swap_phase_shift(0), m_x_swap_amplitude(0),
+  m_x_swap_amplitude_shift(0), m_x_move_amplitude(0), m_y_swap_phase_shift(0), m_y_swap_amplitude(
+    0), m_y_swap_amplitude_shift(0), m_y_move_amplitude(0), m_y_move_amplitude_shift(0),
+  m_z_swap_phase_shift(0), m_z_swap_amplitude(0), m_z_swap_amplitude_shift(0), m_z_move_amplitude(
+    0), m_z_move_amplitude_Goal(0), m_z_move_amplitude_shift(0), m_a_move_amplitude(0),
+  m_a_move_amplitude_shift(0), m_arm_swing_gain(0), m_arm_roll_Gain(0), hip_comp(0.0), foot_comp(
+    0.0), time_unit(8.0), m_x_move_phase_shift(0.5_pi), m_x_move_amplitude_shift(0),
+  m_y_move_phase_shift(0.5_pi), m_z_move_phase_shift(0.5_pi), m_a_move_phase_shift(0.5_pi),
+  m_ctrl_running(false), m_real_running(false), m_time(0), x_move(0.0), y_move(0.0), a_move(0.0),
+  a_move_aim_on(false), is_compute_odometry(false)
+{
+  reset_angles();
+}
+
+void Kinematic::reset_angles()
+{
+  for (auto & angle : joint_angles) {
+    angle = 0.0;
+  }
+}
+
+const std::array<double, 18> & Kinematic::get_angles() const
+{
+  return joint_angles;
+}
+
+void Kinematic::set_running_state(bool running_state)
+{
+  m_ctrl_running = running_state;
+
+  if (running_state) {
+    m_real_running = running_state;
+  }
+}
+
+bool Kinematic::get_running_state() const
+{
+  return m_ctrl_running;
+}
+
+void Kinematic::set_move_amplitude(double x, double y, double a, bool aim_on)
+{
+  x_move = x;
+  y_move = y;
+  a_move = a;
+  a_move_aim_on = aim_on;
+}
+
+double Kinematic::get_x_move_amplitude() const
+{
+  return m_x_move_amplitude;
+}
+
+double Kinematic::get_y_move_amplitude() const
+{
+  return m_y_move_amplitude;
+}
+
+double Kinematic::get_hip_offest() const
+{
+  return hip_pitch_offset + hip_comp;
+}
+
+bool Kinematic::time_to_compute_odometry() const
+{
+  return is_compute_odometry;
+}
+
+void Kinematic::stop_kinematic()
+{
+  m_ctrl_running = false;
+  m_real_running = false;
+
+  m_x_move_amplitude = 0.0;
+  m_y_move_amplitude = 0.0;
+  m_a_move_amplitude = 0.0;
+  m_z_move_amplitude = 0.0;
+
+  update_times();
+  update_move_amplitude();
+}
+
+double Kinematic::wsin(
+  double time, double period, double period_shift, double mag,
+  double mag_shift) const
+{
+  return mag * sin(2_pi / period * time - period_shift) + mag_shift;
+}
+
+bool Kinematic::compute_inverse_kinematic(
+  std::string leg, double x, double y, double z, double a,
+  double b, double c)
+{
+  size_t index = (leg == "right") ? 0 : 6;
+
+  {
+    using keisan::Point3;
+    using keisan::Euler;
+    using keisan::Matrix;
+
+    auto matrix_translation = keisan::translation_matrix(Point3(x, y, z - leg_length));
+    auto matrix_rotation = keisan::rotation_matrix(
+      Euler(
+        keisan::make_radian(a),
+        keisan::make_radian(b), keisan::make_radian(c)));
```

Comment:
seems the `x, y, z` and `a, b, c` is only used here, i suggest we could just use `keisan::Point3` and `keisan::Euler` as parameters in this function (we could later compensate leg length by just calling `point.z -= leg_length`).

Reply:
done